lib/libutee/tee_api.c

```diff
@@ -245,6 +248,136 @@ TEE_Result TEE_InvokeTACommand(TEE_TASessionHandle session,
 	return res;
 }
 
+#ifdef CFG_OCALL
+TEE_Result TEE_InvokeCACommand(uint32_t cancellationRequestTimeout,
+			       uint32_t commandID, uint32_t paramTypes,
+			       TEE_Param params[TEE_NUM_PARAMS],
+			       uint32_t *returnOrigin)
+{
+	const TEE_UUID uuid = PTA_SYSTEM_UUID;
+	TEE_Result res = TEE_SUCCESS;
+	uint32_t ret_origin = TEE_ORIGIN_TEE;
+	struct utee_params pta_up = { };
+	struct utee_params ocall_up = { };
+	void *buffer = NULL;
+	size_t size = 0;
+	size_t n = 0;
+
+	/* Open session with the System PTA, if necessary */
+	if (!__tee_api_system_session) {
+		res = TEE_OpenTASession(&uuid, TEE_TIMEOUT_INFINITE, 0, NULL,
+					&__tee_api_system_session, &ret_origin);
+		/* The System PTA is optional */
+		if (res == TEE_ERROR_ITEM_NOT_FOUND &&
+		    ret_origin == TEE_ORIGIN_TEE) {
+			res = TEE_ERROR_NOT_SUPPORTED;
+			ret_origin = TEE_ORIGIN_API;
+		}
+		if (res)
+			goto exit;
+	}
+
+	/* Convert the OCALL's parameters into a utee_params structure */
+	ocall_up.types = paramTypes;
+	for (n = 0; n < TEE_NUM_PARAMS; n++) {
+		switch (TEE_PARAM_TYPE_GET(paramTypes, n)) {
+		case TEE_PARAM_TYPE_VALUE_INPUT:
+		case TEE_PARAM_TYPE_VALUE_INOUT:
+			ocall_up.vals[n * 2] = params[n].value.a;
+			ocall_up.vals[n * 2 + 1] = params[n].value.b;
+			break;
+		case TEE_PARAM_TYPE_MEMREF_INPUT:
+		case TEE_PARAM_TYPE_MEMREF_OUTPUT:
+		case TEE_PARAM_TYPE_MEMREF_INOUT:
+			buffer = params[n].memref.buffer;
+			size = params[n].memref.size;
+			if ((buffer && !size) || (!buffer && size)) {
+				res = TEE_ERROR_BAD_PARAMETERS;
+				ret_origin = TEE_ORIGIN_API;
+				goto exit;
+			}
+			ocall_up.vals[n * 2] = (vaddr_t)buffer;
+			ocall_up.vals[n * 2 + 1] = size;
+			break;
+		default:
+			break;
+		}
+	}
+
+	/* Construct the parameters for the call to the PTA */
+	pta_up.vals[0] = commandID;
+	pta_up.vals[2] = (uintptr_t)&ocall_up;
+	pta_up.vals[3] = sizeof(ocall_up);
+	pta_up.types = TEE_PARAM_TYPES(TEE_PARAM_TYPE_VALUE_INOUT,
+				       TEE_PARAM_TYPE_MEMREF_INOUT,
+				       TEE_PARAM_TYPE_NONE,
+				       TEE_PARAM_TYPE_NONE);
+
+	/* Send the OCALL request */
+	res = _utee_invoke_ta_command((uintptr_t)__tee_api_system_session,
+				      cancellationRequestTimeout,
+				      PTA_SYSTEM_OCALL,
+				      &pta_up,
+				      &ret_origin);
```

Comment:
I considered funneling this call via a function in `lib/libutee/tee_system_pta.c` which would eventually land in `invoke_system_pta` and reuse the static `sess` variable to manage the session to the system PTA.

However, this function calls `TEE_InvokeTACommand`, which performs parameter processing and creates copies of any `TEE_PARAM_TYPE_MEMREF_*` buffers to isolate the caller TA's memory for the TA being called. This is unnecessary for OCALLs since the callee is the system PTA which can read the caller's memory anyway. To avoid this copy, I construct the `struct utee_params` for the call manually.

If I were to create a wrapper for this call in `lib/libutee/tee_system_pta.c`, I'd have to either create the call's `struct utee_params` in `Invoke_CACommand` and pass that to the wrapper, which would require exposing `struct utee_params` in `tee_internal_api_extensions.h`, or move the entire logic of `Invoke_CACommand` to `lib/libutee/tee_system_pta.c`. This would yield a new function with the same signature as `Invoke_CACommand`, which doesn't seem clean to me.

Reply:
I see, so it makes sense to keep most of the processing here.